crates/biome_js_analyze/src/lint/nursery/use_iterable_callback_return.rs

```diff
@@ -0,0 +1,376 @@
+        recommended: false,
+    }
+}
+
+impl Rule for UseIterableCallbackReturn {
+    type Query = ControlFlowGraph;
+    type State = RuleState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let cfg = ctx.query();
+
+        if !JsFunctionExpression::can_cast(cfg.node.kind())
+            && !JsArrowFunctionExpression::can_cast(cfg.node.kind())
+        {
+            return None;
+        }
+
+        let parent_node = cfg.node.parent()?;
+        let argument_list = JsCallArgumentList::cast_ref(&parent_node)?;
+        let call_expression = argument_list
+            .parent::<JsCallArguments>()?
+            .parent::<JsCallExpression>()?;
+
+        let callee = call_expression.callee().ok()?;
+
+        let member_expression = callee.as_js_static_member_expression()?;
+        let member = member_expression.member().ok()?;
+        let member_name = member.as_js_name()?;
+        let member_name = member_name.value_token().ok()?;
+
+        let method_config = ITERABLE_METHOD_INFOS.get(member_name.text_trimmed())?;
+
+        let arg_position = argument_list
+            .elements()
+            .position(|arg| arg.node.is_ok() && arg.node.unwrap().syntax().eq(&cfg.node))?;
+
+        if arg_position != method_config.callback_argument_position {
+            return None;
+        }
+
+        if let Some(global_name) = method_config.global_name {
+            let (_, name) = global_identifier(&member_expression.object().ok()?)?;
+            if name.text() != global_name {
+                return None;
+            }
+        }
+
+        let returns_info = get_function_returns_info(cfg);
+
+        let mut problems: Vec<RuleProblemKind> = Vec::new();
+        let member_range = member_expression.member().ok()?.range();
+        if method_config.return_value_required {
+            if returns_info.has_paths_without_returns {
+                if returns_info.returns_with_value.is_empty() {
+                    problems.push(RuleProblemKind::MissingReturn);
+                } else {
+                    problems.push(RuleProblemKind::NotAllPathsReturnValue);
+                }
+            }
+            for return_range in returns_info.returns_without_value {
+                problems.push(RuleProblemKind::UnexpectedEmptyReturn(return_range));
+            }
+        } else {
+            for return_range in returns_info.returns_with_value {
+                problems.push(RuleProblemKind::UnexpectedReturnWithValue(return_range));
+            }
+        }
+
+        if !problems.is_empty() {
+            Some(RuleState {
+                problems: problems.into_boxed_slice(),
+                callee_range: member_range,
+                method_info: method_config,
+            })
+        } else {
+            None
+        }
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
```

Comment:
As part of our rule pillars, we try to have diagnostics tell people what to do to resolve the issue.

https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#what-a-rule-should-say-to-the-user